lib/reverseproxy/revers_proxy.go

```diff
@@ -0,0 +1,125 @@
+package reverseproxy
+
+import (
+	"flag"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
+)
+
+const (
+	targetURL     = "vm-target-url"
+	forwardHeader = "X-Forwarded-For"
+)
+
+var (
+	maxIdleConnsPerBackend = flag.Int("maxIdleConnsPerBackend", 100, "The maximum number of idle connections vmauth can open per each backend host")
+)
+
+// ReversProxy represents simple revers proxy based on http.Client
+type ReversProxy struct {
+	client *http.Client
+}
+
+// New initialize revers proxy
+func New() *ReversProxy {
+	tr := prepareTransport()
+	return &ReversProxy{
+		client: &http.Client{
+			Transport: tr,
+		},
+	}
+}
+
+// ServeHTTP serve http requests
+func (rr *ReversProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	rr.handle(w, r)
+}
+
+// handle works with income request, update it to outcome and copy response to the requester
+func (rr *ReversProxy) handle(w http.ResponseWriter, r *http.Request) {
+	tURL, err := getTargetURL(r)
+	if err != nil {
+		logger.Panicf("BUG: unexpected error when parsing targetURL=%q: %s", targetURL, err)
+	}
+
+	ctx := r.Context()
+	proxyReq := r.Clone(ctx)
+
+	proxyReq.URL = tURL
+	if r.ContentLength == 0 {
+		proxyReq.Body = nil // Issue 16036: nil Body for http.Transport retries
+	}
+	if proxyReq.Body != nil {
+		defer func() { _ = proxyReq.Body.Close() }()
+	}
+
+	if err := updateXForwardHeader(r.RemoteAddr, proxyReq.Header); err != nil {
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+		return
+	}
+
+	resp, err := rr.client.Transport.RoundTrip(proxyReq)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+		return
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	for k, vv := range resp.Header {
+		for _, v := range vv {
+			w.Header().Set(k, v)
+		}
+	}
+
+	w.WriteHeader(resp.StatusCode)
+	_, err = io.Copy(w, resp.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+		return
+	}
+}
+
+// prepareTransport builds http transport which can be configurable
+func prepareTransport() *http.Transport {
+	tr := &http.Transport{
+		DisableCompression: true,
+		// Disable HTTP/2.0, since VictoriaMetrics components don't support HTTP/2.0 (because there is no sense in this).
+		ForceAttemptHTTP2: false,
+	}
+
+	tr.MaxIdleConnsPerHost = *maxIdleConnsPerBackend
+	if tr.MaxIdleConns != 0 && tr.MaxIdleConns < tr.MaxIdleConnsPerHost {
```

Comment:
Why do we need to check `MaxIdleConns` for transport that has just been created, won't it always be 0?

Reply:
fixed